### ISSUE TITLE
Graphic artifacts (small gray squares) on the desktop when the window is minimized

### ISSUE THREAD

**Issue by thomaslevesque** [microsoft/terminal#15219]
### Windows Terminal version

1.16.10262.0

### Windows build number

10.0.22621.1555

### Steps to reproduce

I'm not sure exactly when it happens. I use WT without issues, and after a while I notice the glitch; not sure what makes it appear. Once it starts appearing, it stays. 

### Expected Behavior

There should be no graphic artifacts on the desktop when I minimize or restore the window.

### Actual Behavior

Whenever I minimize the window, small gray squares appear just above the taskbar:

![image](https://user-images.githubusercontent.com/1864281/233515606-58c1fbfa-72b8-47ee-ae81-b9b29d481764.png)

Their number seems to match the number of open tabs in WT, although it's a bit inconsistent.

When I restore the window (not maximized), I see just one small gray square in the top left corner of the screen:

![image](https://user-images.githubusercontent.com/1864281/233515748-438ef995-351d-4725-b375-4c5b97e5e3c0.png)

The gray squares are only visible on the desktop. If any window masks that part of the desktop, they don't appear (in other words, they appear below all windows, with the lowest z-index).

**Comment by DHowett** [comment]
Hey, do you happen to have "mica4all" or "mica everywhere" (I can't remember the name!) installed?

**Comment by thomaslevesque** [comment]
@DHowett no, never heard of it.

**Comment by thomaslevesque** [comment]
Not exactly. I just see small grey squares, not rectangles like in your screenshot. Might be related, though.
(haven't seen this in a while btw, maybe it's been fixed?)

**Comment by ClaireCJS** [comment]
> Not exactly. I just see small grey squares, not rectangles like in your screenshot. Might be related, though. (haven't seen this in a while btw, maybe it's been fixed?)

Interesting. If you look at my white rectangles, the right edge of them is a grey square - so it seems related, but different?

**Comment by zadjii-msft** [comment]
If someone's still seeing this regularly, it might be helpful to know what the actual window class for that window is (with like, spyxx or winspy). I'd guess that it's either:

* `PseudoConsoleWindow`
* `DesktopWindowXamlSource`

and which one it is would really change my theory of the root cause. 

Also, knowing what version you're on would also help. If it's a xaml island window (for some reason), then 1.18 probably changed the way this presents itself.

**Comment by zadjii-msft** [comment]
A thought: why are we using `WS_OVERLAPPEDWINDOW` instead of `WS_OVERLAPPED`..?

**Comment by ClaireCJS** [comment]
> If someone's still seeing this regularly

I am. I hate it so much and it plagues me every day, but it's better than not using Windows Terminal and not even being able to see filenames with emoji & unicode in them. I hate hate hate hate hate hate hate it.




>it might be helpful to know what the actual window class for that window is (with like, spyxx or winspy). 

Okay! It seems Winspy is discontinued and now there's something called WinSpy++ that's basically the same thing.

I'd guess that it's either:
> 
> * `PseudoConsoleWindow`
> * `DesktopWindowXamlSource`
> 
> and which one it is would really change my theory of the root cause.

The classname is actually SHELLDLL_DefView.  Seems to be associated with...explorer.exe?!?!?!

None of this happened for years (30yrs of that command line, 2 yrs of that command line on the same machine) until I installed Windows Terminal, then it happened, and always happened, consistently... Even if I ran my TCC/TakeCommand (jpsoft.com) command line without Windows Terminal wrapped around it.  I'm pretty sure at some point I even tried uninstalling Windows Terminal, but nope, so I reinstalled it. It's quite perplexing and I'd love to get to the bottom of it.


> 
> Also, knowing what version you're on would also help. If it's a xaml island window (for some reason), then 1.18 probably changed the way this presents itself.

Which version of windows terminal?  1.17.11461.0
I'm really limited by what's in the app store. I've failed to build it on my own and decided that's not an endeavor suited to me. However, I'd happily try a precompiled build of a later version if someone would supply it to me.   [Windows 10.0.19045.3393]

Here's more info. I really wish WinSpy++ had a text-dump functionality to avoid awkward screenshotting:

![image](https://github.com/microsoft/terminal/assets/789591/71a9bd58-d45e-4faa-8b42-6d97ad3f428c)

![image](https://github.com/microsoft/terminal/assets/789591/686332a8-e6f5-4cb1-9c5a-fcbce386e49b)

![image](https://github.com/microsoft/terminal/assets/789591/13f9ef39-4b56-4bcd-99fe-dd738d233529)

![image](https://github.com/microsoft/terminal/assets/789591/84f01f21-4022-45ab-8d0d-8f627312249c)

![image](https://github.com/microsoft/terminal/assets/789591/f4cb612a-10ca-4d12-b05b-3c2236c986cd)

![image](https://github.com/microsoft/terminal/assets/789591/8ab38da2-f7a7-4e72-975d-efe30200cb3b)

**Comment by zadjii-msft** [comment]
> SHELLDLL_DefView

what the fishsticks is that? I've literally never heard of that, and it's for sure not something the Terminal is making ourselves.

**Comment by zadjii-msft** [comment]
Wait hold up - what's the class of the teeny tiny window: 
![image](https://github.com/microsoft/terminal/assets/18356694/7a48a88f-25ea-4af1-adf2-2ce771fd03a2)

not the one that's seemingly appearing by the taskbar. There's... a lot of shell customizations you got going on there. I wouldn't be surprised if it was some weird combo of all those running together. 

But that tiny white rectangle - that's the one I am pretty sure is caused by us

**Comment by ClaireCJS** [comment]
There's no customizations I weren't running for at least 10 years before the day I installed windows terminal, and minimizing and un-minimizing windows terminal windows are what cause these to display. 

This does not happen on any of the other machines where I use the exact same customizations without windows terminal.

Turning displayfusion off (which adds some buttons on window taskbars) does not change this.

**Comment by ClaireCJS** [comment]
Anyway, those screenshots _were_ from the tiny black square that isn't near the taskbar.

I'll happily gather screenshots of the ones by the taskbar too, if that helps!

**Comment by zadjii-msft** [comment]
what the fuck

![minimized-phwnd](https://github.com/microsoft/terminal/assets/18356694/dfbf8193-ad60-4567-a387-97ccbdf43327)

okay, so there's a good chance the `PseudoConsoleWindow` is moving there when it's minimized. _why tho_

**Comment by ClaireCJS** [comment]
Is this a eureka moment?   There's a lot I understand, but there's more I don't understand, including a lot of what's going on here! :)

**Comment by zadjii-msft** [comment]
not really. If you've got winspy++ screenshots of what that rectangle looks like on your machine, that'll be helpful. I can't seem to find a combo of flags to turn on to make it become visible[^1] on my machine. 


[^1]: I mean, yea the window is WS_VISIBLE, but I mean like, visible to a human / rendered to the screen. Maybe this is some weird transparency thing?

**Comment by ClaireCJS** [comment]
I can do that.

Well, they look like this:

![image](https://github.com/microsoft/terminal/assets/789591/3e4f0dc6-8915-4559-8b48-366a618b1968)


[winspy screenshots coming next comment]

**Comment by ClaireCJS** [comment]
Here's the Winspy++ screenshots:
![image](https://github.com/microsoft/terminal/assets/789591/0367987c-4cec-4759-8531-be9d071fa608)

![image](https://github.com/microsoft/terminal/assets/789591/53554513-4dd3-4b89-b03f-eb983cd9825c)

![image](https://github.com/microsoft/terminal/assets/789591/8587e61c-1572-4b94-96a8-b9675b0cd970)

![image](https://github.com/microsoft/terminal/assets/789591/2f19a3fe-6ea6-4622-9d46-0a3ee25979b9)

![image](https://github.com/microsoft/terminal/assets/789591/32396f6c-cba8-4353-8751-f81b929e7bd2)
![image](https://github.com/microsoft/terminal/assets/789591/d5439c7b-dcd0-4485-8464-e81ad0da643f)

**Comment by zadjii-msft** [comment]
Well, a couple curious things there:
* There's no owner, so this isn't the conpty window. That's clear from the Class not being PseudoConsoleWindow, but that makes me think it's unrelated to us. 
* It's 3072x1728 px big? That's not the window we're looking for, that's for sure. It might be the _desktop_? 
  * Actually, looks like the first screenshots you shared were for _the same window handle_. 
  
Looks like it was a miss on the finder, for whatever reason.

Here's an idea though: if you go into WinSpy++, then hit `More >>`, you should be able to find a `cmd.exe` (or `tcc.exe` or whatever the shell is), with a pseudoconsolewindow underneath it (after double-clicking). Hit the <kbd>Autoupdate</kbd> button, and see how _that_ window reacts to the Terminal restoring and being minimized. The  <kbd>Flash</kbd> button is also helpful for ID'ing where on the screen a window should be (it will flash the bounds of a window even if the window isn't visible). 

Once you've got a lock on the right pseudoconsolewindow, share those screenshots?

**Comment by ClaireCJS** [comment]
3072x1728? That's... not even the resolution of the desktop i'm dropping the finder on, it's a 1920x1080 desktop. 3 out of 4 of mine are. It's probably some combination of my 4 displays (3*1080p + 1*VGA). 

Yes, it appears that it's missing actually. And I missed that it was missing. oof

**Comment by ClaireCJS** [comment]
> Here's an idea though: if you go into WinSpy++, then hit `More >>`, you should be able to find a `cmd.exe` (or `tcc.exe` or whatever the shell is), with a pseudoconsolewindow underneath it (after double-clicking). Hit the Autoupdate button, and see how _that_ window reacts to the Terminal restoring and being minimized. The Flash button is also helpful for ID'ing where on the screen a window should be (it will flash the bounds of a window even if the window isn't visible).
> 
> Once you've got a lock on the right pseudoconsolewindow, share those screenshots?

Well, it's kind of hard to share the screen of the flash function flashing, but basically...

![image](https://github.com/microsoft/terminal/assets/789591/7b2dabd8-3878-4da3-a3c8-5d13a4414188)

1) Flashing the entire TCC flashes a square with the size and placement that is approximately the default size and placement TCC comes up with, when it starts on my screen. And it always starts on the same monitor, even if my taskbar shortcuts are on different monitors. TCC before Windows Terminal when launched directly did not behave that way. 

I'm starting to think Windows Terminal and multiple monitors and moving TCC to a different monitor somehow is causing some sort of disconnect where TCC might have leftover aspects of its windows in the original monitor from where it was moved from (and originally opened into) (i.e. primary monitor).

2) Flashing 002416B0 TCC.EXE "TCC:c658" captured the upper-left square and about 3 more square's width to the right. Small rectangle.

3) Flashing 07442Dc2 PsuedoConsoleWindow captured the upper-left square and a few pixels to the left of it.

4) Flashing Default IME - i see nothing

I'm staring to think this is specific to how Windows Terminal interacts with TCC when TCC is moved to another monitor. I have thought it behaved strange, but not quite been able to put my finger on it.

**Comment by zadjii-msft** [comment]
Okay, maybe we've been tying two related but ultimately separate things here. 

Quick way to separate these out: @ClaireCJS Do you see the same behavior without any `tcc.exe` tabs? Just a plain-old PowerShell or Command Prompt tab. Let's eliminate one variable here. 

---

We definitely do have reports of the tiny, top-left square _without_ TCC, but this may be exacerbated by it. I'm tempted to switch to 

```diff
diff --git a/src/interactivity/base/InteractivityFactory.cpp b/src/interactivity/base/InteractivityFactory.cpp
index fe8fc55cd..4e675cb85 100644
--- a/src/interactivity/base/InteractivityFactory.cpp
+++ b/src/interactivity/base/InteractivityFactory.cpp
@@ -330,8 +330,8 @@ using namespace Microsoft::Console::Interactivity;
                 // will return the console handle again, not the owning
                 // terminal's handle. It's not entirely clear why, but WS_POPUP
                 // is absolutely vital for this to work correctly.
-                const auto windowStyle = WS_OVERLAPPEDWINDOW | WS_POPUP;
-                const auto exStyles = WS_EX_TOOLWINDOW | WS_EX_TRANSPARENT | WS_EX_LAYERED | WS_EX_NOACTIVATE;
+                const auto windowStyle = WS_OVERLAPPED | WS_POPUP;^M
+                const auto exStyles = WS_EX_TRANSPARENT | WS_EX_LAYERED | WS_EX_NOACTIVATE;^M

                 // Attempt to create window.
                 hwnd = CreateWindowExW(exStyles,
```

But I need to sync with my user32 contact before I'd feel comfortable with that. It _seems_ to remove the size when it's restored (the tiny rectangle), but not the tiny one....

**Comment by ClaireCJS** [comment]
Oh, it's definitely TCC+Windows Terminal specific. I think it was phrased that way on it's own issue but then my issue got merged/mentioned into this issue since they're the same thing with apparently two different causes.

But yea, to answer your question, it doesn't happen with powershell tabs

but it also doesn't happen when TCC is run without windows terminal

I suspect TCC is doing something weird and nonstandard and Windows Terminal isn't detecting something it meant to detect correctly. It's kind of been the broad pattern.

**Comment by zadjii-msft** [comment]
Oh butts you're right, there's a whole other thread for that. 

I'm gonna de-merge them, move relevant comments over, and collapse things in this thread. Thanks for bearing with me ☺️

**Comment by zadjii-msft** [comment]
How did I miss

> (haven't seen this in a while btw, maybe it's been fixed?)

Well then. I'm going to tenuously mark this `Fixed?`, and give the community till the end of 1.20 to find a repro. If we don't hear anything, I'm gonna go ahead and say this was magically fixed along the way. 

If anyone DOES end up with a repro for this _specifically_[^1], then a [Task Manager dump](https://github.com/microsoft/terminal/wiki/Toubleshooting-Tips#capture-with-task-manager) of _OpenConsole.exe_ (NOT WindowsTerminal.exe) might be enlightening. 

[^1]: not the TCC thing tracked in #15536

**Comment by thomaslevesque** [comment]
Sounds good to me. If I see it again, I'll let you know.

**Comment by ShakoFarhad** [comment]
This happens concistently still for me. I have reformatted and reinstalled windows 11 several times (Windows 11 Pro) on a modern and powerful laptop from Lenovo. I did not experience this on my previous laptop which was a Windows Surface Book 2.

I have noticed these squares appear in the same way as described above after WSL ubuntu terminal being open over night or just over a long period of time. I noticed that the squares dissappear after restarting the PC, and once I noticed the squares dissappeared after running "wsl" in powershell as it seemed like WSL had turned itself off over night.

If possible, tell me if there is anything I can do to fix this.

**Comment by zadjii-msft** [comment]
@ShakoFarhad Sorry I missed that! You should be able to share that with the team following these steps: https://github.com/microsoft/terminal/wiki/Troubleshooting-Tips#share-with-the-team

Is it a dump of `windowsterminal.exe`, or `openconsole.exe`? We're pretty confident the squares are due to `openconsole.exe`, (which Terminal uses)

**Comment by thomaslevesque** [comment]
Seeing this again, for the first time in a while:
![image](https://github.com/microsoft/terminal/assets/1864281/f3991f14-625a-4b88-8e9a-a777ac3ca4b8)

> If anyone DOES end up with a repro for this _specifically_[1](#user-content-fn-1-2065bcfdb585b21ceba7415dd9d3c31d), then a [Task Manager dump](https://github.com/microsoft/terminal/wiki/Troubleshooting-Tips#capture-with-task-manager) of _OpenConsole.exe_ (NOT WindowsTerminal.exe) might be enlightening.

Which one? I have 4 instances of OpenConsole.exe

**Comment by jonasmarco** [comment]
![image](https://github.com/user-attachments/assets/dab6dd56-a286-4ea3-9c3c-313ab3b592e0)

07/31/2024

**Comment by zadjii-msft** [comment]
I'm pretty sure that #17828 wasn't ultimately related to this, but can someone try out the latest Terminal Canary (>=1.23.2421.0) and see if that fixed this? 

You can grab terminal canary from [aka.ms/terminal-canary-installer](https://aka.ms/terminal-canary-installer)

**Comment by DHowett** [comment]
> same here 28/10/2024

Are you using Preview or Stable?

**Comment by Lukiticas** [comment]
> > same here 28/10/2024
> 
> Are you using Preview or Stable? Which version _exactly_?

sorry for the late response, Stable, version 1.21.2911.0, win11 23H2 latest patch

**Comment by kovalexal** [comment]
Still happening for me in Stable, version 1.21.3231.0, win11 24H2. Are there any resolutions that might work (at least as a temporary solution)?

It's a pity to change a terminal application because of a single bug.
